xarray/core/dataset.py

```diff
@@ -1123,7 +1123,13 @@ def _copy_listed(self, names: Iterable[Hashable]) -> "Dataset":
         coord_names = set()
         indexes: Dict[Hashable, pd.Index] = {}
 
-        for name in names:
+        def key(name):
+            try:
+                return list(self._variables.keys()).index(name)
+            except ValueError:
+                return len(self._variables)
+
+        for name in sorted(names, key=key):
```

Comment:
I don't think we need to sort names here. They are explicitly provided by users so we can copy whatever they use.

Reply:
that's true. However, we then need to sort in `DatasetCoordinates.to_dataset`, because that passes `_coord_names`.